src/runtime/callbacks.ts

```diff
@@ -0,0 +1,271 @@
+  ): void {
+    this.setCallback_(CallbackId.PAY_CONFIRM_OPENED, callback as Callback);
+  }
+
+  /**
+   * @return Whether the callback has been found.
+   */
+  triggerPayConfirmOpened(activityIframeView: ActivityIframeView): boolean {
+    return this.trigger_(CallbackId.PAY_CONFIRM_OPENED, activityIframeView);
+  }
+
+  setOnSubscribeRequest(callback: () => void): void {
+    this.setCallback_(CallbackId.SUBSCRIBE_REQUEST, callback);
+  }
+
+  /**
+   * @return Whether the callback has been found.
+   */
+  triggerSubscribeRequest(): boolean {
+    return this.trigger_(CallbackId.SUBSCRIBE_REQUEST, true);
+  }
+
+  hasSubscribeRequestCallback(): boolean {
+    return !!this.callbacks_[CallbackId.SUBSCRIBE_REQUEST];
+  }
+
+  setOnOffersFlowRequest(callback: () => void): void {
+    this.setCallback_(CallbackId.OFFERS_FLOW_REQUEST, callback);
+  }
+
+  /**
+   * @return Whether the callback has been found.
+   */
+  triggerOffersFlowRequest(): boolean {
+    return this.trigger_(CallbackId.OFFERS_FLOW_REQUEST, true);
+  }
+
+  hasOffersFlowRequestCallback(): boolean {
+    return !!this.callbacks_[CallbackId.OFFERS_FLOW_REQUEST];
+  }
+
+  setOnSubscribeResponse(
+    callback: (responsePromise: Promise<SubscribeResponse>) => void
+  ): void {
+    warn(
+      `[swg.js:setOnSubscribeResponse]: This method has been deprecated, please switch usages to 'setOnPaymentResponse'`
+    );
+    this.setCallback_(CallbackId.PAYMENT_RESPONSE, callback as Callback);
+  }
+
+  setOnContributionResponse(
+    callback: (responsePromise: Promise<SubscribeResponse>) => void
+  ): void {
+    warn(
+      `[swg.js:setOnContributionResponse]: This method has been deprecated, please switch usages to 'setOnPaymentResponse'`
+    );
+    this.setCallback_(CallbackId.PAYMENT_RESPONSE, callback as Callback);
+  }
+
+  setOnPaymentResponse(
+    callback: (responsePromise: Promise<SubscribeResponse>) => void
+  ): void {
+    this.setCallback_(CallbackId.PAYMENT_RESPONSE, callback as Callback);
+  }
+
+  /**
+   * @return Whether the callback has been found.
+   */
+  triggerPaymentResponse(responsePromise: Promise<SubscribeResponse>): boolean {
+    this.paymentResponsePromise_ = responsePromise.then(
+      (res) => {
+        this.trigger_(
+          CallbackId.PAYMENT_RESPONSE,
+          Promise.resolve(res.clone())
+        );
+      },
+      (reason) => {
+        if (isCancelError(reason)) {
+          return;
+        }
+        throw reason;
+      }
+    );
+    return !!this.callbacks_[CallbackId.PAYMENT_RESPONSE];
+  }
+
+  hasPaymentResponsePending(): boolean {
+    return !!this.resultBuffer_[CallbackId.PAYMENT_RESPONSE];
+  }
+
+  setOnFlowStarted(
+    callback: ({flow, data}: {flow: string; data: unknown}) => void
+  ): void {
+    this.setCallback_(CallbackId.FLOW_STARTED, callback as Callback);
+  }
+
+  /**
+   * @return Whether the callback has been found.
+   */
+  triggerFlowStarted(flow: string, data = {}): boolean {
```

Comment:
similarly here, not sure if type for `data` needs to be explicitly defined.

Reply:
In this case, the default value implies `{}`. What do you think?
<img width="567" alt="image" src="https://user-images.githubusercontent.com/1216762/231295327-ae973228-4c42-4654-ab60-f687d83d8475.png">

Reply:
```
Object has similarities to both {} and object, but is slightly looser. Use {} for a type that include everything except null and undefined
```

Seems good to me!